### PR TITLE
add typescript definitions

### LINKED ISSUE #345: include typescript definitions in npm package
@FezVrasta would you consider publishing a `.d.ts` file in the NPM package for typescript consumers? [`@types/popper.js`](https://www.npmjs.com/package/@types/popper.js) is already a thing, but including types in the source package itself provides a more consistent experience. it also allows libraries to depend on your interfaces without introducing an `@types` dependency that is unnecessary for JS consumers.  

i've been improving the `@types` definitions and could be ready with a PR today or tomorrow?

### PR DESCRIPTION
**Pull request by giladgray**
#### Fixes #345 

- add `index.d.ts` to popper package (and `types` entry in package.json)
- add a note in PULL_REQUEST_TEMPLATE to remind folks to keep this updated

### CODE REVIEW

**Comment by giladgray** [comment]
@FezVrasta is there any work required to ensure the new `index.d.ts` file gets published to NPM? also, are you okay with placing it in the package root?

**Comment by giladgray** [comment]
@FezVrasta wow that CLA-assistant wants a lot of permissions! Why does it need to write to my public repos?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1,9 +1,11 @@
 <!--
-Thanks for your interest in contributing to Popper.js!
```
Comment (FezVrasta) [inline .github/PULL_REQUEST_TEMPLATE.md]: Please don't remove the trailing whitespaces, they are meaningful in markdown

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
-Thanks for your interest in contributing to Popper.js!  
-Please, make sure to fulfill the following conditions before submitting your Pull Request.  
+Thanks for your interest in contributing to Popper.js!
+Please, make sure to fulfill the following conditions before submitting your Pull Request.
 
 ### Make sure to
 
 1. Make sure the tests are passing, you can check it running `npm test` with Google Chrome installed;
 2. Add any relevant test to cover the code you have changed and/or added;
+3. If you change the public API, update the Typescript definitions accordingly:
```
Comment (FezVrasta) [inline .github/PULL_REQUEST_TEMPLATE.md]: Didn't you decide to maintain them yourself? If this is still the case is rephrase this to a "nice to have but not requiree action"

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (giladgray) [inline .github/PULL_REQUEST_TEMPLATE.md]: my editor did that. but also this is a comment, so they're really not meaningful here: it is invisible in rendered output. 

i much prefer using blank lines to create paragraphs in markdown (instead of the awkward trailing two spaces) because i always have my editor set to trim trailing whitespace.

Comment (giladgray) [inline .github/PULL_REQUEST_TEMPLATE.md]: also blank lines are visible in source so i can tell immediately that you meant to have two paragraphs.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (giladgray) [inline .github/PULL_REQUEST_TEMPLATE.md]: i'll take responsibility for this file and be the first line of defense should something go wrong, but i think it's worth adding a comment like this so folks know it exists. 

if someone forgets, or has trouble editing it, ping me for help. i'll amend the note to this effect.

**Comment by giladgray** [comment]
@FezVrasta added a note about myself in PR template. let me know what else you'd like to see here. would be great to merge and ship a new release soon so I can consume it in Blueprint!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (FezVrasta) [inline .github/PULL_REQUEST_TEMPLATE.md]: As stated previously, I don't want to lose contributions because someone gets scared reading he needs to know TypeScript to contribute.

If the point says "If you change the public API, it would be nice to try to update the TypeScript definitions accordingly. This is not required but will help a lot." (or something similar) then I'd be happy to merge this.

Comment (giladgray) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd be wary of contributors changing the public API at will--probably not likely to happen often, but when it does then special attention will be essential anyway.

i'll update the message to be more accomodating

**Comment by FezVrasta** [comment]
ok it looks great now. I'm not sure how this new file you added gets read by TypeScript but:

1. Right now it will not be included in the npm package
2. If people import from `dist/esm/popper.js` should the new file be in that folder as well with the same name?
3. Same of point 2 applies for the umd build and the next/es build on the root directory

**Comment by giladgray** [comment]
The `types` entry in package.json tells TS compiler where to find typings, that's the only thing that matters.
Fair to add `index.d.ts` to files list then?

**Comment by FezVrasta** [comment]
Yes sure

**Comment by FezVrasta** [comment]
released as 1.11.0
